Add Concurrency Basics section

Avasil opened the pull request:
I've written something regarding Concurrency and I hope it could be good addition to documentation. :)

My idea is to gather all relevant information that is usually scattered around the internet, books and gitter conversations. With your review/contributions I believe it will become decent reference point for understanding basics of concurrency that are relevant to `Cats-Effect` and others.

As time passes and I understand more I'll try adding more information - I want to add short "Stack safety" section next but I scrapped it for now to do this PR and release anything. 

If you like this initiative we could make a TODO list and have corresponding issues to gradually make it more comprehensive.

It's probably the first time I'm writing something that long in English so I will really appreciate any suggestions regarding wording, style etc. so the message is clear. Also most of the content is my own understanding so it might be completely wrong. :P

Avasil commented on site/src/main/tut/concurrency/basics.md:
I would like to have it in first version but I haven't been able to find any decent sources about it in contest of `Future`, `IO`, `Task` etc. 

Does anybody have explanation what are benefits of fork join pool (which is the default) vs regular fixed thread pool? How is it possible to split `IO`?

ericluria commented on site/src/main/tut/concurrency/basics.md:
Per discussion in the Gitter channel: I think it might also help to clarify the purpose of the `implicit ec: ExecutionContext` parameter of `infiniteIO`. As far as I understand, it's used by `repeat.start`. So when you call `prog.unsafeRunSync()`, what's happening with `ec` while `ecOne` and `ecTwo` do all the work?

alexandru replied:
@Avasil the API on master is going to change thanks to #289 — you should not require an `ExecutionContext` here, but a `ContextShift`.

gatorcse commented on site/src/main/tut/concurrency/basics.md:
Awkwardly worded. Potential fix: _If it is time to change threads, the currently running thread is **preemted** - it saves its state and the context switch happens._

Avasil replied:
Thank you @gatorcse !

alexandru commented:
@Avasil I'm striving to get all PRs for RC3 merged, so if there's anything else you want to do with this document, today would be a good time, otherwise I'll merge as is.

The website doesn't have the same requirements as code, so no pressure, we can merge or do other corrections later.

alexandru commented on site/src/main/tut/concurrency/basics.md:
Here too we need `ContextShift`, not `ExecutionContext`.

alexandru commented on site/src/main/tut/concurrency/basics.md:
Given you've given `shift` examples in this document, we'll need a sample for `ContextShift#evalOn` here, since it's meant for blocking operations.

Avasil commented:
Thanks @alexandru - when I was writing it there wasn't #289 yet, I can update it later today (probably in the evening)

alexandru commented:
@Avasil #289 is now on `master`.

As I said, documentation doesn't have the same requirements, so no rush, but would be cool to merge this too this week.

Zelenya commented on site/src/main/tut/concurrency/basics.md:
Should the **green threads** be explained somewhere above? Like after **Threading (on JVM)**?

Zelenya commented on site/src/main/tut/concurrency/basics.md:
```diff
@@ -0,0 +1,253 @@
+Before any thread can start doing real work the OS needs to store state of earlier task and restore the state 
+for the current one. This cleanup has actually nontrivial cost so from performance point of view 
+the most efficient situation for CPU-bound tasks is when we execute as many threads as the number of available 
+operating systems' threads favoring synchronous execution whenever possible.
```
This sentence seems really complex to me. Is it possible to rephrase or split?

Avasil commented:
@alexandru
I've updated examples for `RC3` and addressed all current suggestions. :)

From my side it's good to merge (but I'm still curious about Fork Join Pool for `IO` :P) if you skimmed through it and didn't notice any glaring errors.

Avasil commented:
Thanks for corrections @rossabaker :)

I think I'll make an issue with todo list and maybe it will be good place to give any criticism
